src/crimson/osd/recovery_backend.h

```diff
@@ -178,8 +190,9 @@ class RecoveryBackend {
   };
   using RecoveryBlockingEvent =
     crimson::AggregateBlockingEvent<WaitForObjectRecovery::BlockingEvent>;
+  using WaitForObjectRecoveryRef = boost::intrusive_ptr<WaitForObjectRecovery>;
 protected:
-  std::map<hobject_t, WaitForObjectRecovery> recovering;
+  std::map<hobject_t, WaitForObjectRecoveryRef> recovering;
```

Comment:
Yeah, the blocker must not die before the blocking is complete. This wasn't ensured because of:

```cpp
void PGRecovery::on_global_recover (
  const hobject_t& soid,
  const object_stat_sum_t& stat_diff,
  const bool is_delete)
{ 
  logger().info("{} {}", __func__, soid);
  pg->get_peering_state().object_recovered(soid, stat_diff);
  // ...
  pg->get_recovery_backend()->remove_recovering(soid);
}
```

```cpp
  void remove_recovering(const hobject_t& soid) {
    recovering.erase(soid);
  }
```